tidy3d/components/data/dataset.py

```diff
@@ -1678,3 +2474,69 @@ def sel(
 
 
 UnstructuredGridDatasetType = Union[TriangularGridDataset, TetrahedralGridDataset]
+CustomSpatialDataType = Union[SpatialDataArray, TriangularGridDataset, TetrahedralGridDataset]
+CustomSpatialDataTypeAnnotated = Union[
+    SpatialDataArray, annotate_type(Union[TriangularGridDataset, TetrahedralGridDataset])
+]
+
+
+def _get_numpy_array(data_array: Union[ArrayLike, DataArray, UnstructuredGridDataset]) -> ArrayLike:
+    """Get numpy representation of dataarray/dataset values."""
+    if isinstance(data_array, UnstructuredGridDataset):
+        return data_array.values.values
+    if isinstance(data_array, xr.DataArray):
+        return data_array.values
+    return np.array(data_array)
+
+
+def _zeros_like(
+    data_array: Union[ArrayLike, xr.DataArray, UnstructuredGridDataset]
+) -> Union[ArrayLike, xr.DataArray, UnstructuredGridDataset]:
+    """Get a zeroed replica of dataarray/dataset."""
+    if isinstance(data_array, UnstructuredGridDataset):
+        return data_array.updated_copy(values=xr.zeros_like(data_array.values))
+    if isinstance(data_array, xr.DataArray):
+        return xr.zeros_like(data_array)
+    return np.zeros_like(data_array)
+
+
+def _ones_like(
+    data_array: Union[ArrayLike, xr.DataArray, UnstructuredGridDataset]
+) -> Union[ArrayLike, xr.DataArray, UnstructuredGridDataset]:
+    """Get a unity replica of dataarray/dataset."""
+    if isinstance(data_array, UnstructuredGridDataset):
+        return data_array.updated_copy(values=xr.ones_like(data_array.values))
+    if isinstance(data_array, xr.DataArray):
+        return xr.ones_like(data_array)
+    return np.ones_like(data_array)
+
+
+def _check_same_coordinates(
```

Comment:
Are we consistent in the tolerances, like using `rel_tol=1e-10` above but the `np.allclose` defaults here?